hamlet/celery.py

```diff
@@ -300,13 +302,68 @@ def ml_subject_assistant_export_to_microsoft(
     service.
     """
     
+    try:
+        export = MLSubjectAssistantExport.objects.get(pk=export_id)
+        target_filename = 'ml-subject-assistant-{}-export{}.json'.format(
+            export.subject_set_id,
+            export.id,
+        )
+        source_filepath = ''
+        
+        export.status = MLSubjectAssistantExport.RUNNING
+        export.save()
+
+        # Get the Subjects data
+        data = ml_subject_assistant_export_to_microsoft_pt1_get_subjects_data(export_id, access_token)
+
+        # Create the file to be exported
+        source_filepath = ml_subject_assistant_export_to_microsoft_pt2_create_file(export_id, data, target_filename)
+
+        # Upload the file to Azure, and get a shareable URL to the file
+        shareable_file_url = ml_subject_assistant_export_to_microsoft_pt3_create_shareable_azure_blob(source_filepath, target_filename)
+        
+        # SUCCESS
+        export.status = MLSubjectAssistantExport.COMPLETE
+        
+        # Save the created file to the database
+        # NOTE: this is technically optional, and only used as a backup
+        with open(source_filepath, 'rb') as out_f:
```

Comment:
Looks like you're opening the file readonly before trying to write to it?

Reply:
I'm not sure what I'm doing wrong here, and would appreciate some clarification.

Here's my thought process:

1. I'm using `write_subject_set_export()` as a template, so...

```
# In write_subject_set_export()...
# A. I understand that this block creates a temporary file in the filesystem
with tempfile.NamedTemporaryFile(
    'w+',
    encoding='utf-8',
    dir=settings.TMP_STORAGE_PATH,
    delete=False,
) as out_f:
    ...

# B. I understand that this block reads that temporary file, and saves it to the database.
with open(out_f_name, 'rb') as out_f:
    export.csv.save(
        ...
    )
export.status = SubjectSetExport.COMPLETE
export.save()
```

2. In `ml_subject_assistant_export_to_microsoft()`, I similarly created...

```
# A. this part creates a temporary file in the filesystem
source_filepath = ml_subject_assistant_export_to_microsoft_pt2_create_file(export_id, data, target_filename)

...

# B. this block reads that temporary file, and saves it to the database.
with open(source_filepath, 'rb') as out_f:
    export.json.save(
        ...
    )
...
export.save()
```

After I create the temporary file, I don't think I'm writing to it any further? (`source_filepath` is the equivalent to `out_f` in your code, and points to the temporarily created file... I can change this to `out_f` if it's less confusing to follow a standard, or a more declarative `path_of_temporary_file`)

Reply:
Ah, OK – it was the `out_f` name that was confusing me, so I guess I'm getting confused by my own bad naming scheme from my code :)